File: solutions/python/circular-buffer/1/circular_buffer.py
class BufferFullException(BufferError):
    """Exception raised when CircularBuffer is full.

    message: explanation of the error.

    """
    def __init__(self, message):
        pass


class BufferEmptyException(BufferError):
    """Exception raised when CircularBuffer is empty.

    message: explanation of the error.

    """
    def __init__(self, message):
        pass


class CircularBuffer:
    def __init__(self, capacity):
        self.size = 0
        self.items = [None] * capacity
        self.capacity = capacity

        self.read_pointer = 0
        self.write_pointer = 0


    def read(self):
        if self.size == 0:
            raise BufferEmptyException("Circular buffer is empty")

        self.size -= 1

        item = self.items[self.read_pointer]
        self.__increment_read_pointer()

        return item


    def write(self, data):
        if self.size == self.capacity:
            raise BufferFullException("Circular buffer is full")

        self.size += 1

        self.items[self.write_pointer] = data
        self.__increment_write_pointer()


    def overwrite(self, data):
        if self.size == self.capacity:
            self.size -= 1
            self.__increment_read_pointer()

        self.write(data)


    def clear(self):
        self.size = 0
        self.read_pointer = self.write_pointer


    def __increment_pointer(self, position):
        return (position + 1) % self.capacity

    def __increment_read_pointer(self):
        self.read_pointer = self.__increment_pointer(self.read_pointer)

    def __increment_write_pointer(self):
        self.write_pointer = self.__increment_pointer(self.write_pointer)

File: solutions/csharp/circular-buffer/1/CircularBuffer.cs
using System;

public class CircularBuffer<T>
{
    private T[] buffer;

    private int readPointer = 0;
    private int writePointer = 0;
    
    private int size = 0;
    
    public CircularBuffer(int capacity)
    {
        buffer = new T[capacity];
    }

    public T Read()
    {
        if (IsEmpty())
            throw new InvalidOperationException("Buffer is empty.");

        var result = buffer[readPointer];

        AdvanceReadPointer();
        size--;
        
        return result;
    }

    public void Write(T value)
    {
        if (IsFull())
            throw new InvalidOperationException("Buffer is full.");

        buffer[writePointer] = value;
        
        AdvanceWritePointer();
        size++;
    }
    
    public void Overwrite(T value)
    {
        buffer[writePointer] = value;

        AdvanceWritePointer();

        if (IsFull()) 
            AdvanceReadPointer();
        else
            size++;
    }

    public void Clear()
    {
        readPointer = writePointer;
        size = 0;
    }

    private void AdvanceReadPointer()
    {
        readPointer++;
        if (readPointer == buffer.Length)
           readPointer = 0;
    }

    private void AdvanceWritePointer()
    {
        writePointer++;
        if (writePointer == buffer.Length)
           writePointer = 0;
    }


    private bool IsEmpty() => size == 0;
    
    private bool IsFull() => size == buffer.Length;
}